pkg/chunk/aws/s3_storage_client.go

```diff
@@ -160,3 +141,64 @@ func (a s3ObjectClient) bucketFromKey(key string) string {
 
 	return a.bucketNames[hash%uint32(len(a.bucketNames))]
 }
+
+// Get object from the store
+func (a S3ObjectClient) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, error) {
+	var resp *s3.GetObjectOutput
+
+	// Map the key into a bucket
+	bucket := a.bucketFromKey(objectKey)
+
+	err := instrument.CollectedRequest(ctx, "S3.GetObject", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		var err error
+		resp, err = a.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(objectKey),
+		})
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// Put object into the store
+func (a S3ObjectClient) PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error {
+	return instrument.CollectedRequest(ctx, "S3.Put", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		_, err := a.S3.PutObjectWithContext(ctx, &s3.PutObjectInput{
+			Body:   object,
+			Bucket: aws.String(a.bucketFromKey(objectKey)),
+			Key:    aws.String(objectKey),
+		})
+		return err
+	})
+}
+
+// List objects from the store
+func (a S3ObjectClient) List(ctx context.Context, prefix string) (map[string]time.Time, error) {
+	objectKeysWithMtime := map[string]time.Time{}
+	prefixWithSep := prefix + "/"
+
+	for i := range a.bucketNames {
+		err := instrument.CollectedRequest(ctx, "S3.List", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+			output, err := a.S3.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{Bucket: &a.bucketNames[i], Prefix: &prefix})
+			if err != nil {
+				return err
+			}
+
+			for i := range output.Contents {
```

Comment:
The Thanos implementation uses the Minio client, which internally does paginate across results. It may be a good reference to checkout to see how it works, even if in our case we don't need to work with channels like the Minio client does.